### DISCUSSION TITLE
Uploaded Video previews correctly, playback is upside down

### DISCUSSION THREAD

**Discussion by matteohoeren** [photoprism/photoprism#4619] [Q&A]
#### 1. What is not working as documented?

When i playback a video (.MOV Format, HEVC H.265, Recorded on an iPhone 13 Mini) it is flipped 180° degrees. The preview is displayed correctly. It's kinda weird because it happens on nearly all said videos, but not every single one of them. Their orientation is right on my iPhone and my Photos Library on Mac. 

#### 2. How can we reproduce it?

Upload a Video with Photosync from an iPhone, filmed in portrait orientation.

#### 3. What behavior do you expect?

Playback the video with the right orientation.

#### 4. What could be the cause of your problem?

Photoprism infers the orientation? 

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

Here are 2 screenshots:

https://ibb.co/25dDyqq
https://ibb.co/xscpGpk

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: AMD64, Version 24.04, Build 240915-e1280b2fb (docker on linux)

(b) Database Type & Version: MariaDB

(c) Operating System Types & Versions: MacOS

(d) Browser Types & Versions: Chrome-based browser

(e) Ad Blockers, Browser Plugins, and/or Firewall Software? None

#### 7. On what kind of device is PhotoPrism installed?

(a) Device / Processor Type: Intel N100

(b) Physical Memory & Swap Space in GB: 32GB RAM

(c) Storage Type: SSD 2TB

(d) Anything else that might be helpful to know?

#### 8. Do you use a Reverse Proxy, Firewall, VPN, or CDN?

No

**Comment by graciousgrey** [comment]
Do you have hardware transcoding enabled?

If yes this might be related to: 
- https://github.com/photoprism/photoprism/issues/4403

**Comment by matteohoeren** [comment]
@graciousgrey No, i don't have that configured in my docker-compose so it defaults to software transcoding. Should i try enabling hardware transcoding?

**Comment by lastzero** [comment]
Before you try anything else, we recommend that you check the metadata of both files with Exiftool, as described here in our documentation:
- https://docs.photoprism.app/getting-started/troubleshooting/metadata/

**Comment by lastzero** [comment]
I will move this to discussions while the issue is still being investigated.

**Comment by matteohoeren** [comment]
So, here's the output of exiftool, with one of my **problematic** (showing Rotation: 90) videos:

```json
[{
  "SourceFile": "Matteo/Zuletzt/IMG_6917.MOV",
  "ExifToolVersion": 12.76,
  "FileName": "IMG_6917.MOV",
  "Directory": "Matteo/Zuletzt",
  "FileSize": 12624748,
  "FileModifyDate": "2024:11:03 15:43:24+01:00",
  "FileAccessDate": "2024:11:06 12:23:46+01:00",
  "FileInodeChangeDate": "2024:11:04 15:54:28+01:00",
  "FilePermissions": 100664,
  "FileType": "MOV",
  "FileTypeExtension": "MOV",
  "MIMEType": "video/quicktime",
  "MajorBrand": "qt  ",
  "MinorVersion": "0.0.0",
  "CompatibleBrands": ["qt  "],
  "MediaDataSize": 12610624,
  "MediaDataOffset": 36,
  "MovieHeaderVersion": 0,
  "CreateDate": "2024:11:03 14:43:24",
  "ModifyDate": "2024:11:03 14:43:31",
  "TimeScale": 600,
  "Duration": 6.565,
  "PreferredRate": 1,
  "PreferredVolume": 1,
  "PreviewTime": 0,
  "PreviewDuration": 0,
  "PosterTime": 0,
  "SelectionTime": 0,
  "SelectionDuration": 0,
  "CurrentTime": 0,
  "NextTrackID": 8,
  "TrackHeaderVersion": 0,
  "TrackCreateDate": "2024:11:03 14:43:24",
  "TrackModifyDate": "2024:11:03 14:43:31",
  "TrackID": 1,
  "TrackDuration": 6.565,
  "TrackLayer": 0,
  "TrackVolume": 0,
  "ImageWidth": 1920,
  "ImageHeight": 1080,
  "CleanApertureDimensions": "1920 1080",
  "ProductionApertureDimensions": "1920 1080",
  "EncodedPixelsDimensions": "1920 1080",
  "GraphicsMode": 64,
  "OpColor": "32768 32768 32768",
  "CompressorID": "hvc1",
  "SourceImageWidth": 1920,
  "SourceImageHeight": 1080,
  "XResolution": 72,
  "YResolution": 72,
  "CompressorName": "HEVC",
  "BitDepth": 24,
  "VideoFrameRate": 30,
  "CameraLensModel-und-DE": "iPhone 13 mini back camera 5.1mm f/1.6",
  "CameraFocalLength35mmEquivalent-und-DE": 27,
  "Balance": 0,
  "AudioFormat": "mp4a",
  "AudioChannels": 2,
  "AudioBitsPerSample": 16,
  "AudioSampleRate": 44100,
  "PurchaseFileFormat": "mp4a",
  "Warning": "[minor] The ExtractEmbedded option may find more tags in the media data",
  "ContentDescribes": 1,
  "MatrixStructure": "1 0 0 0 1 0 0 0 1",
  "MediaHeaderVersion": 0,
  "MediaCreateDate": "2024:11:03 14:43:24",
  "MediaModifyDate": "2024:11:03 14:43:31",
  "MediaTimeScale": 600,
  "MediaDuration": 6.56666666666667,
  "MediaLanguageCode": "und",
  "GenMediaVersion": 0,
  "GenFlags": "0 0 0",
  "GenGraphicsMode": 64,
  "GenOpColor": "32768 32768 32768",
  "GenBalance": 0,
  "HandlerClass": "dhlr",
  "HandlerVendorID": "appl",
  "HandlerDescription": "Core Media Data Handler",
  "MetaFormat": "mebx",
  "HandlerType": "mdta",
  "Full-frame-rate-playback-intent": 1,
  "Make": "Apple",
  "Model": "iPhone 13 mini",
  "Software": "18.0.1",
  "CreationDate": "2024:11:03 15:43:24+01:00",
  "CameraLensModel": "iPhone 13 mini back camera 5.1mm f/1.6",
  "CameraFocalLength35mmEquivalent": 27,
  "ImageSize": "1920 1080",
  "Megapixels": 2.0736,
  "AvgBitrate": 15367097,
  "Rotation": 90
}]
```

Here is an output from an **unproblematic** (showing rotation: 0) file:

```json
[{
  "SourceFile": "Matteo/Zuletzt/IMG_6873.MP4",
  "ExifToolVersion": 12.76,
  "FileName": "IMG_6873.MP4",
  "Directory": "Matteo/Zuletzt",
  "FileSize": 10271989,
  "FileModifyDate": "2024:10:21 17:52:17+02:00",
  "FileAccessDate": "2024:11:06 12:26:23+01:00",
  "FileInodeChangeDate": "2024:11:04 15:54:56+01:00",
  "FilePermissions": 100664,
  "FileType": "MP4",
  "FileTypeExtension": "MP4",
  "MIMEType": "video/mp4",
  "MajorBrand": "mp42",
  "MinorVersion": "0.0.1",
  "CompatibleBrands": ["isom","mp41","mp42"],
  "MediaDataSize": 10264183,
  "MediaDataOffset": 44,
  "MovieHeaderVersion": 0,
  "CreateDate": "2024:10:21 15:52:07",
  "ModifyDate": "2024:10:21 15:52:17",
  "TimeScale": 44100,
  "Duration": 10.3650113378685,
  "PreferredRate": 1,
  "PreferredVolume": 1,
  "PreviewTime": 0,
  "PreviewDuration": 0,
  "PosterTime": 0,
  "SelectionTime": 0,
  "SelectionDuration": 0,
  "CurrentTime": 0,
  "NextTrackID": 3,
  "TrackHeaderVersion": 0,
  "TrackCreateDate": "2024:10:21 15:52:07",
  "TrackModifyDate": "2024:10:21 15:52:17",
  "TrackID": 1,
  "TrackDuration": 10.3540362811791,
  "TrackLayer": 0,
  "TrackVolume": 1,
  "Balance": 0,
  "AudioFormat": "mp4a",
  "AudioChannels": 2,
  "AudioBitsPerSample": 16,
  "AudioSampleRate": 44100,
  "MatrixStructure": "1 0 0 0 1 0 0 0 1",
  "ImageWidth": 1080,
  "ImageHeight": 1920,
  "MediaHeaderVersion": 0,
  "MediaCreateDate": "2024:10:21 15:52:07",
  "MediaModifyDate": "2024:10:21 15:52:17",
  "MediaTimeScale": 600,
  "MediaDuration": 10.365,
  "MediaLanguageCode": "und",
  "HandlerType": "vide",
  "HandlerDescription": "Core Media Video",
  "GraphicsMode": 0,
  "OpColor": "0 0 0",
  "CompressorID": "hvc1",
  "SourceImageWidth": 1080,
  "SourceImageHeight": 1920,
  "XResolution": 72,
  "YResolution": 72,
  "BitDepth": 24,
  "ColorProfiles": "nclx",
  "ColorPrimaries": 1,
  "TransferCharacteristics": 1,
  "MatrixCoefficients": 1,
  "VideoFullRangeFlag": 0,
  "VideoFrameRate": 30.0048239266763,
  "ImageSize": "1080 1920",
  "Megapixels": 2.0736,
  "AvgBitrate": 7922178,
  "Rotation": 0
}]
```

Can you guys make sense of that? I have no idea why some files my be tagged with a Rotation and some without a rotation. I think my quick-fix would be so set all Rotation: 90 entries to 0 again, should to the trick i think.

**Reply by ffaraone** [reply]
I'm experiencing the same issue, for just some of the videos the preview is ok but when you reproduce it, it appear rotated 180 degrees. If you download it, it is ok.

**Comment by haozhun** [comment]
All vertical videos (and live photos) look ok in preview, but are rotated 180 degrees when I try to play them. They are shot on iPhone 13 Pro, 16, 16 Pro (.mov H265). 

A few important notes:
*  I have this problem in desktop Chrome only (Mac). But the videos play fine in Safari (Mac or iOS). They also play fine if I download them and play them in QuickTime, VLC, IINA. 
* All of these played fine in the past in PhotoPrism. Therefore, it's likely related to some recent changes. I'm running PhotoPrism 240915 (:latest), and haven't upgraded for a while. Therefore, it's likely not PhotoPrism version. Maybe it's Chrome version?

I checked the exif data. All the vertical videos have

```JSON
  "Rotation": 90,
```

There is no entry for `Orientation`.

cc @lastzero (I feel that I've provided important additional details to this issue. Therefore, I hope to get your attention.)

**Reply by lastzero** [reply]
Do you use software or hardware video transcoding?

**Reply by haozhun** [reply]
Hardware.

In Chrome on Mac (Apple silicone or Intel silicone), if I disable "use graphics acceleration when available", PhotoPrism would do server side transcoding (for H265 videos). If PhotoPrism does server-side transcoding, the video would display correctly (not flipped).

**Reply by haozhun** [reply]
Safari on iPhone and Safari on Mac are also using hardware decoding.

**Reply by lastzero** [reply]
So you did not (originally) [transcode the affected video](https://docs.photoprism.app/getting-started/advanced/transcoding/), and the problem **only occurs** when you play the video directly in your web browser? Is that correct?

In this case, the browser should display the videos as intended, because without transcoding (touching the original file), I don't know how to change its behavior? If you have other information and know how to do it, please let us know! :)

**Reply by srett** [reply]
@lastzero I just investigated that issue a bit and found that (part of) the solution seems to be what the user in https://github.com/photoprism/photoprism/issues/4403#issuecomment-2226927540 also found - adding `-noautorotate` before the input argument.
The reason according to some googling around seems to be that `ffmpeg` [by default reads the display rotation](https://ffmpeg.org/ffmpeg.html#Video-Options) (see `-autorotate`) from the video metadata, then automatically applies the rotation during filtering/transcoding, and *excludes that metadata* from the output file, to prevent the playback software from applying the rotation again during playback.  
The issue now appears to be that with hardware encoding, the rotation is *not* automatically applied, I guess because it had to be done in hardware too and either ffmpeg doesn't pass that information along, or because it gets lost somewhere along the way. However, ffmpeg still strips the rotation information from the metadata, so the output file will have the same orientation as the input, but as the metadata is lost, the playback software (browser) will not apply that rotation itself.  
By passing `-noautorotate` you instruct ffmpeg to not do any rotation (which it doesn't do anyways with hw encode) and instead leave the metadata intact while copying it to the output file via `-map_metadata 0`. So the video gets rotated in the browser and everything is fine.  

But then why does it happen for some people who do not use hardware encoding, and only with some browsers, and only sometimes?

If you keep the rotation metadata in the file, then instead of applying the rotation during transcode, you require the playback software (browser) to honor the rotation information when displaying the video. This very evidently doesn't appear do be the case in some setups, as @haozhun reports for example on MacOS with Chrome when using hardware decode. I assume that just like with encoding via ffmpeg and hw, in that case the hw pipeline simply ignores the rotation information and displays the video as is. When disabling hw acceleration in the browser, the decode takes another path where the rotation information is correctly handled. (Note: I think when they said *"if I disable "use graphics acceleration when available", PhotoPrism would do server side transcoding"* that's not really whats happening, or does PhotoPrism decide whether to transcode or not depending on browser settings/capabilities?).

So, for browsers that honor the rotation metadata, adding `-noautorotate` appears to be the simplest solution. It could even be applied for software transcode to make switching between hw and sw decode less surprising regarding the output.

Please also note that this problem will also occur for videos that PhotoPrism will not transcode at all (since they don't exceed configured codec, size and bitrate requirements), since those might very well also contain rotation information, and get sent to the browser unprocessed.

So to "fix" this problem reliably for every possible platform and browser you'd either need to:
- Force transcoding of *every* video that has rotation metadata while making sure ffmpeg will correctly rotate the video during transcode and *remove* the metadata - this needs to be somehow implemented for hw encode, somewhere in ffmpeg/vaapi/qsv/???
- Variation of this: When using hw encode, determine the rotation metadata before transcode and manually add some rotation magic to -vf, assuming this can be sone in hw at all
- Make sure every browser implements proper handling of rotation metadata and just go with `-noautorotate`
- Tell users that some browsers just suck (Chrome with hw decode) and still go with `-noautorotate` :wink:

**Reply by haozhun** [reply]
Thank you very much. I don't totally understand it (little understanding of video processing), but I hope that the information that you provided will be helpful toward a resolution.

For now, I've been using Safari (instead of Chrome) whenever accessing PhotoPrism, as a workaround.

> Note: I think when they said "if I disable "use graphics acceleration when available", PhotoPrism would do server side transcoding" that's not really whats happening, or does PhotoPrism decide whether to transcode or not depending on browser settings/capabilities?

Yes, PhotoPrism decide whether to transcode based on browser capability. In Chrome on MacOS, if I one disables "use graphics acceleration when available", Chrome will report that it doesn't support H265.

**Reply by haozhun** [reply]
@srett. In my case, all vertical videos are displayed 180 degrees off. The video's metadata suggests 90 degree rotation. From my naive understanding, that means the number of rotations is off by two, not one.

**Reply by haozhun** [reply]
I decided to do some simple investigation myself. First, I'll say that this is indeed a problem with Chrome on Mac, not a PhotoPrism specific problem.

I loaded the following page in Chrome on Mac and in Safari on Mac. The video is a vertical video shot on iPhone. In Safari, it displays correctly. On Chrome, the rotation is 180 degrees off.

```
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <title>Simple Video Page</title>
</head>
<body>
  <video width="640" height="360" controls>
    <source src="video.mov" type="video/mp4">
  </video>
</body>
</html>
```

**Reply by srett** [reply]
> @srett. In my case, all vertical videos are displayed 180 degrees off. The video's metadata suggests 90 degree rotation. From my naive understanding, that means the number of rotations is off by two, not one.

Huh, that sounds like yet another issue then, and not what I saw in my limited testing, or at least a different manifestation of the same underlying problem.

> Yes, PhotoPrism decide whether to transcode based on browser capability. In Chrome on MacOS, if I one disables "use graphics acceleration when available", Chrome will report that it doesn't support H265.

Interesting, maybe I'm not up to date here, but in the past I'm pretty sure that PP would transcode everything that isn't H264, unless you completely disable transcode in the settings, but I didn't verify just now as I was mostly focused on the specific issue with hardware accelerated transcode.

**Reply by trexx** [reply]
@srett Thanks so much for the detailed breakdown. As someone who uses QuickSync to encode videos on my N100, I noticed videos weren't correctly rotated after transcoding on both my Android phone and Chromium on Linux. Adding -noautorotate is a temporary workaround for my issue for my browser / device combination as it seems the video rotation metadata is respected from where I view it.
